### ISSUE TITLE
TextField cursor is misplaced when regain focus & textAlign is end, right, or center 

### ISSUE THREAD

**Issue by Aniki312** [flutter/flutter#21679]
Another issue similar to issue #18483

After input some text, and click the left side of the textfield
the cursor is still at the end of the textfield, while the actual focus is at the start of the textfield.
You can see from the gif how new characters are appended, and also how backspace stop working after additional characters are deleted. 

![textfield_cursor_issue](https://user-images.githubusercontent.com/2377511/45285888-e58e0d00-b516-11e8-8d0e-44178c19dd7d.gif)

```dart
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(),
      home: new Scaffold(
        appBar: new AppBar(
          title: new Text("TextField Cursor"),
        ),
        body: new Center(
          child: TextField(
            textAlign: TextAlign.end,
          ),
        ),
      ),
    );
  }
}
```

Flutter doctor -v 
```
[✓] Flutter (Channel master, v0.8.2-pre.24, on Mac OS X 10.13.6 17G65, locale zh-Hant-TW)
    • Flutter version 0.8.2-pre.24 at /Users/aniki/git/flutter
    • Framework revision 4d9bbc10f2 (26 hours ago), 2018-09-09 03:15:34 -0400
    • Engine revision 1770f88548
    • Dart version 2.1.0-dev.4.0.flutter-cd9a42239f

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/aniki/Library/Android/sdk
    • Android NDK at /Users/aniki/Library/Android/sdk/ndk-bundle
    • Platform android-28, build-tools 27.0.3
    • ANDROID_HOME = /Users/aniki/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.4.1, Build version 9F2000
    • ios-deploy 1.9.2
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 28.0.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] VS Code (version 1.27.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.18.0

[✓] Connected devices (2 available)
    • Android SDK built for x86 64 • emulator-5554                        • android-x64 • Android 9 (API 28) (emulator)
    • iPhone X                     • BE0CF3F0-CCBB-469B-A1FE-BE5530372455 • ios         • iOS 11.4 (simulator)
```

**Comment by ozc** [comment]
I have the same issue with RTL languages. I will add flutter doctor output.

**Comment by DhudumVishal** [comment]
I have the same issue, the focus stays on correct position but the cursor is shown on the right corner.

**Comment by p30arena** [comment]
Same Issue when Locale is set to a RTL Language;
When using LTR characters "only" or "combined" with RTL characters in the TextField.

```
[√] Flutter (Channel dev, v0.11.2, on Microsoft Windows [Version 6.3.9600], locale en-US)
    • Flutter version 0.11.2 at M:\flutter
    • Framework revision e32cd85446 (2 weeks ago), 2018-11-09 14:26:21 -0800
    • Engine revision 316f6365a6
    • Dart version 2.1.0 (build 2.1.0-dev.9.3 9c07fb64c4)

[√] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at D:\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • ANDROID_HOME = D:\Android\sdk
    • Java binary at: C:\Program Files\Java\jdk1.8.0_45\bin\java
    • Java version Java(TM) SE Runtime Environment (build 1.8.0_45-b15)
    • All Android licenses accepted.

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 6.0 (API 23) (emulator)

! Doctor found issues in 1 category.
```

**Comment by p30arena** [comment]
I can't also adjust the Cursor between characters or at the beginning.

**Comment by Kartik1607** [comment]
+1 Having same issue with center text. 

```
TextField(
    textAlign: TextAlign.center,
    maxLines: 1,
    decoration: InputDecoration(
        border: InputBorder.none,
        prefixIcon: Icon(Icons.list, color: Colors.black54,),
        suffixIcon: Icon(Icons.search, color: Colors.black54,),
        hintText: "Search Restaurant",
    ),
)
```
![Image](https://i.imgur.com/7sTcLN1.gif)

**Comment by p30arena** [comment]
@zoechi 
This issue is severe and needs special attention.
Why not label it as "BUG"?

**Comment by p30arena** [comment]
https://github.com/flutter/flutter/issues/21679#issuecomment-441585623

problem still persists in flutter 1.1.2

https://gfycat.com/CoolGrandioseBlueshark
![coolgrandioseblueshark](https://user-images.githubusercontent.com/1261332/50142105-79ae0400-02be-11e9-909f-bb6b6fe79a26.gif)

```
[√] Flutter (Channel dev, v1.1.2, on Microsoft Windows [Version 6.3.9600], locale en-US)
    • Flutter version 1.1.2 at M:\flutter
    • Framework revision d675bbbd3a (4 days ago), 2018-12-14 10:59:31 -0500
    • Engine revision 101b27da7a
    • Dart version 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)

[√] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at D:\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • ANDROID_HOME = D:\Android\sdk
    • Java binary at: C:\Program Files\Java\jdk1.8.0_45\bin\java
    • Java version Java(TM) SE Runtime Environment (build 1.8.0_45-b15)
    • All Android licenses accepted.

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 6.0 (API 23) (emulator)

! Doctor found issues in 1 category.
```

**Comment by zoechi** [comment]
Duplicate of #21193

**Comment by GaryQian** [comment]
@Kartik1607 Are you still finding that the caret jumps to the center? That specific behavior should have been fixed by https://github.com/flutter/engine/pull/7164. Have you tried the master channel?

The RTL bug still persists and I am exploring the cause.

**Comment by luizgpa** [comment]
The caret doesn't jump to the center anymore, but it stays in place even if you add more spaces at the end.
![textfield](https://user-images.githubusercontent.com/4613697/50297220-68522c80-0463-11e9-99e7-23c12291a435.gif)
Not the end of the world since I can't think of a legitimate reason for someone doing that.

**Comment by GaryQian** [comment]
I have determined that we should add "fake" metrics to the text when there are extra whitespace at the end of a centered or right aligned text. Currently, we do not allow any trailing whitespace to contribute to the width, and the metrics reflect that. However, this causes unnatural and unintuitive space behavior at the end. We should, for metrics only, include the width caused by whitespace, but not for the layout itself.

**Comment by Kartik1607** [comment]
> @Kartik1607 Are you still finding that the caret jumps to the center? That specific behavior should have been fixed by [flutter/engine#7164](https://github.com/flutter/engine/pull/7164). Have you tried the master channel?
> 
> The RTL bug still persists and I am exploring the cause.

Hi, the caret does not jump when on master channel. Thanks. I'm curious why 'spaces' are not removed from contributing to width?

**Comment by yeganaaa** [comment]
![deepin-screen-recorder_select area_20190111095010](https://user-images.githubusercontent.com/8914248/51013043-557ed380-159b-11e9-953a-a934fe32062e.gif)

**Comment by bigari** [comment]
I've tried to implement a workaround for TextAlign.center or right by placing a TextField inside a sizedBox and changing the content padding dynamically.

Github: https://github.com/bigari/boxed_text_field/blob/master/lib/boxed_text_field.dart

**Comment by DK15** [comment]
I am facing similar issue in `beta` channel wherein the cursor position is at the center of the `hint text` when `textalign` is centered. 
Also, the cursor doesn't move to the starting position of the entered text and gets stuck at the position after first letter, as shown below:

[Cursor position issue](https://media.giphy.com/media/iOAjqjmvzXrAaljdOk/giphy.gif)

**Comment by 1abzar** [comment]
what a big bug !! hope google team solve this problem

**Comment by zoechi** [comment]
@DK15 @1abzar please check `dev` channel. I can't reproduce with

```
[✓] Flutter (Channel master, v1.2.1-pre.117, on Mac OS X 10.14.3 18D42, locale en-AT)
```

I'm going to close. Please add a comment if you're still seeing this with `dev` channel to have the issue reopened.

**Comment by 1abzar** [comment]
yes I check it again and has problem with `dev` . 
`[√] Flutter (Channel dev, v1.2.0, on Microsoft Windows [Version 10.0.16299.785], locale fa-IR)
`

**Comment by zoechi** [comment]
@1abzar with the code from the initial comment?
If not please provide a minimal but complete runnable reproduction.

**Comment by tje3d** [comment]
@zoechi 
Cursor jump's to first character after typing `ا`, And also jump to middle after push a space.

![ezgif-1-be4fbd0985d1](https://user-images.githubusercontent.com/5238989/52634772-4f9f4980-2edd-11e9-881a-3340149ae892.gif)

**Comment by zoechi** [comment]
@tje3d 
> with the code from the initial comment?
If not please provide a minimal but complete runnable reproduction.

any information about how to reproduce?

**Comment by tje3d** [comment]
@zoechi Direction causes the problem.

Sample code:
```dart
Directionality(
    child: TextFormField(),
    textDirection: TextDirection.rtl,
),
```

flutter doctor:
```
[✓] Flutter (Channel dev, v1.2.0, on Linux, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
```

**Comment by zoechi** [comment]
@tje3d Can you please make a minimal runnable reproduction
as a single file so that we can just copy your code into `lib/main.dart` of a new project and run to reproduce (like the one in the original comment).

**Comment by tje3d** [comment]
@zoechi Here you go. Launch app and press space bar, then look at cursor position.

```dart
import 'package:flutter/material.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'App',
      home: Scaffold(
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Directionality(
                child: TextFormField(
                  initialValue: 'سلام تست',
                  autofocus: true,
                ),
                textDirection: TextDirection.rtl,
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

**Comment by zoechi** [comment]
> Cursor jump's to first character after typing `ا`, And also jump to middle after push a space.

Seems I missed that this only happens when writing special characters.

This is different than the original issue.
Can you please create a new issue with the information you provided above?

**Comment by tje3d** [comment]
@zoechi It's happen with all characters.

Here is what i found:
1. Its happen with all characters
1. Everything works fine with LTR direction
1. If you type `ا` after `ل`, cursor will jumps to begin
1. Cursor always jumps to first lost position when you press space bar.

![ezgif-2-badbf21614b4](https://user-images.githubusercontent.com/5238989/52644021-5fc22380-2ef3-11e9-86a9-05d54ace60cb.gif)

**Comment by zoechi** [comment]
@tje3d please post the output of `flutter doctor -v`

**Comment by tje3d** [comment]
@zoechi 
```
[✓] Flutter (Channel dev, v1.2.0, on Linux, locale en_US.UTF-8)
    • Flutter version 1.2.0 at __HOME__/flutter
    • Framework revision 06b979c4d5 (3 weeks ago), 2019-01-25 14:27:35 -0500
    • Engine revision 36acd02c94
    • Dart version 2.1.1 (build 2.1.1-dev.3.2 f4afaee422)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at __HOME__/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = __HOME__/Android/Sdk
    • Java binary at: __HOME__/Applications/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[!] Android Studio (version 3.3)
    • Android Studio at __HOME__/Applications/android-studio
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] VS Code (version 1.31.0)
    • VS Code at /usr/share/code
    • Flutter extension version 2.22.3

[✓] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 8.1.0 (API 27) (emulator)

! Doctor found issues in 1 category.
```

**Comment by p30arena** [comment]
@zoechi 
I can also confirm that this issue still exists for RTL.

**Comment by 1abzar** [comment]
me 2

> @zoechi
> I can also confirm that this issue still exists for RTL.

me 2

**Comment by zoechi** [comment]
Ok, I'll reopen. 
I was not able to reproduce though.
Any hint what could make the difference would be helpful.

**Comment by p30arena** [comment]
@zoechi 
I can invite you to a remote session so you can test it yourself on my VM.
TeamViewer maybe?

**Comment by zoechi** [comment]
@p30arena I do believe you that you see this behavior, what I tried to figure out is why it does not reproduce here. 
I'll leave it to the engineers who will work on this issue to try to reproduce and ask for more information.

**Comment by GaryQian** [comment]
https://github.com/flutter/flutter/pull/27808 and https://github.com/flutter/engine/pull/7791 have landed and should fix the issue with spaces not appearing when typed at the end. I am looking into the other issues in this issue.

**Comment by GaryQian** [comment]
It looks like the way the arabic ligatures are handled/combined are causing the boundaries of the glyphs in terms of byte offsets to be off. If we ask for a range that does not contain the entire ligature, then we fail to find any bounding boxes, causing the caret to go to the default beginning position.

To solve this, we should find a way to discover which codepoints are being ligatured together and then use that index span to find the bounding box.

**Comment by GaryQian** [comment]
https://github.com/flutter/flutter/issues/22627 is the same as the remaining portions of this bug, although both have good images/gifs so I'll keep both open for now.

**Comment by mansurov1024** [comment]
Hi, is there any progress with this issue?

**Comment by GaryQian** [comment]
This is likely related to https://github.com/flutter/flutter/pull/24797, and the bugs the PR links to.

Not exactly sure how arabic ligatures are encoded, but the lack of zwj characters is the reason why the current iterate through grapheme clusters solution does not work for it.

**Comment by FlutterIODev** [comment]
Are there any workarounds for this problem?

like using text editer from the native platform!

This issue really case a serious problem for us, I hope if we can help to solve it.

**Comment by p30arena** [comment]
@GaryQian 
> 
> 
> This is likely related to #24797, and the bugs the PR links to.
> 
> Not exactly sure how arabic ligatures are encoded, but the lack of zwj characters is the reason why the current iterate through grapheme clusters solution does not work for it.

If you give me some hints, I might be able to help you with Arabic language.

**Comment by abouditube** [comment]
😢😢😢😢😢

**Comment by GaryQian** [comment]
After investigating, a likely cause is that what we call "ghost" runs used to track additional whitespace at the end of lines was appending the extra space to the right instead of the left side of the paragraph.

**Comment by FlutterIODev** [comment]
> After investigating, a likely cause is that what we call "ghost" runs used to track additional whitespace at the end of lines is appending the extra space to the right instead of the left side of the paragraph.

How we can fix this wrong behavior to test if it solves the problem, please ?

**Comment by GaryQian** [comment]
I am working on a fix at the LibTxt layout level. It looks like when a space is entered, it is getting tagged with LTR instead of RTL, which is in turn placing the extra space bounding box at the wrong end of the text.

**Comment by GaryQian** [comment]
https://github.com/flutter/engine/pull/8638 adds manual tracking of text directionality, and infers the 'ghost run' (tracks trailing whitespace) direction from the previous bidi run. This overrides the always-LTR produced by ICU. This in turn allows us to detect and manually place the trailing whitespace box on the correct side of the text for RTL trailing whitespaces.

**Comment by GaryQian** [comment]
So TextField has a parameter called `textDirection`. This should be set to `TextDirection.rtl` for fields primarily meant to take RTL text.

However, in the current implementation, the caret still has the same bug even when rtl is specified. Once this is fixed, the caret should definitely be on the correct side if the textDirection flag is set properly. We are still discussing the proper behavior for the case where the flag is set to LTR and the user types RTL text.

In any case, please set the textDirection to rtl for now, and when the fixes land for this, the spaces should not jump the caret anymore.

**Comment by GaryQian** [comment]
To achieve proper RTL behavior, the `textdirection` should be set to `TextDirection.rtl` or the theme should be set to an RTL theme. Once this is done, we should now handle trailing spaces correctly for rtl and right aligned text.

**Comment by FlutterIODev** [comment]
> To achieve proper RTL behavior, the `textdirection` parameter of TextField should be set to `TextDirection.rtl` or the theme should be set to an RTL theme. Once this is done, we should now handle trailing spaces correctly for rtl and right aligned text.

![Screenshot_٢٠١٩٠٤٢٢-١٨٣٨٥٢](https://user-images.githubusercontent.com/42409569/56509292-4bbc2300-652f-11e9-8d4d-83ad724267ce.jpg)

Thanks, I have downloaded the last version of flutter engine and the problem solved 😍, I need to know if there any previous issue related to that one shown in the picture ( when text selected the selection highlight not aligned) so I can keep tracking it.

**Comment by GaryQian** [comment]
So you should be able to achieve aligned highlights by using BoxHeightStyle.max instead of BoxHeightStyle.tight in the options for GetBoxesForRange. I can double check if there is a direct option for that yet, but in the case it isn't, the call should be present in TextPainter.getBoxesForSelection(). You should be able to extend/override it to use Max bounds instead of Tight.

**Comment by GaryQian** [comment]
See https://github.com/flutter/flutter/issues/31459 for exposing the property.

**Comment by ektelat** [comment]
why this issue closed, the problem with the rtl is still until now

**Comment by GaryQian** [comment]
Are you on the master or dev channels? The fix has been landed on only those branches as of now and will take some time to make it to beta and release. You also need to set the TextDirection to rtl for it to work.

**Comment by tigrenok00** [comment]
@GaryQian 
Hi, is there any possible workaround for now on the stable branch? We cannot possibly go to production with RTL textfields being typed backwards, and I went through the thread but don't understand if a workaround is currently present or not. Just setting TextDirection has no impact on version 1.2.1.
Thanks

**Comment by GaryQian** [comment]
The fix should be already on the dev branch. Any version after https://github.com/flutter/flutter/commit/2e2eeb79de35144edc7665256b8f1097808fd85c should have the fix. This corresponds to release v1.5.5 and any version after it (https://github.com/flutter/flutter/releases/tag/v1.5.5)

As of today, to use this fix, we recommend running `flutter channel dev` to switch to the dev channel and setting TextDirection to rtl.

We release to each of beta and stable on a periodic schedule, and the changes in dev will periodically be promoted to beta (~monthly) and stable (~quarterly) as we see the changes are stable. Dev releases on average about once every 1-3 days, up to a week, and master changes constantly with bleeding edge fixes and features.

Unofficially, I have generally observed that the dev branch (and even master, most of the time) remain fairly stable and any bad releases are marked as bad if any serious problems are found with them. Dev releases must pass a large swath of tests both in the flutter repo itself as well as across Google.

You may attempt to work around this by manually changing your engine.version file to a flutter engine git hash after https://github.com/flutter/engine/commit/ea28bfa13e9d614c126cc3cfe5238c666afdc7d6, however, we cannot guarantee 100% compatibility with mismatched engine and framework versions.

**Comment by tigrenok00** [comment]
@GaryQian Hi! Is there any way of knowing what is the schedule for it to land on the stable branch? Thanks!

**Comment by GaryQian** [comment]
Stable branch roughly releases every quarter, and the last stable release was on May 1st. However it seems that that stable release only included cherrypicked changes and did not happen to pick up the fix for this issue.

I have heard that we are expecting to push another stable release sooner rather than later (shorter interval than a quarter since May), but I don't know what the timeline actually is. It should be along the lines of a 1.7.x release.

**Comment by tigrenok00** [comment]
@GaryQian 
Hi! Did this not made into the new version? Just upgraded to 1.7.8 and set TextDirection RTL and it still reproduces :(

**Comment by GaryQian** [comment]
The fix should have made it in, do you think you can provide a reproduction of this that you are seeing on 1.7.8, and I would be happy to take another look at it.

**Comment by tigrenok00** [comment]
@GaryQian 
Hi! While preparing a reproduction for you, I actually found the cause of the problem.
I'm using the TextField to enter a search term. In the method where it is rendered, I first set the search value for the controller like this:
`_searchTextController.value = new TextEditingValue(
        text: StoreProvider.of<AppState>(context).state.productSearchTerm)`
This is to set the initial value to be the last one entered. 
The productSearchTeam attribute is updated dispatching an action in the `onChange ` method of the TextField.
I am using redux, but it happens with a regular state attribute as well, calling setState from onChange.

If I disable this setting `_searchTextController.value = ` ,  (or typing LTR) everything works fine. However, when using RTL, you see the cursor bug as if you are typing backwards.

**Comment by tigrenok00** [comment]
@GaryQian 
Hi,
It's been some time, but my last comment below is still relevant. Is there a known issue with setting textEditingController value in a stateful widget in RTL mode? Currently direct input in the field works fine, the typing direction is correct. However, setting the value (or text) programmatically in the way I mentioned causes the displayed text to be typed backwards. I do need to set it from code in several situations. Do you have a solution for this?
Thank you!

**Comment by ektelat** [comment]
the same problem is contunie spicially when the text include some sybols or number...
see the video


![problem_flutter](https://user-images.githubusercontent.com/49336458/78810343-ec218680-79d0-11ea-8e94-f2f7b058406b.gif)

**Comment by amirtabasi** [comment]
hi everyone.
excuse me if what i wanna to say is a duplicate ...
in the multiline RTL text when cursor goes to next line, it places at the left side !

![image](https://user-images.githubusercontent.com/55507206/79693365-6da1d000-827f-11ea-9321-709ef2d32f40.png)

**Comment by p30arena** [comment]
I can confirm that there are still issues with Farsi.
